### PR TITLE
IP reservation template

### PR DESCRIPTION
**Pull request by tiadobatima1**

### CODE REVIEW

File: community/cloud-foundation/templates/ip_reservation/ip_reservation.py.schema [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+info:
+  title: IP Reservation
+  author: Sourced Group Inc.
+  description: Reservers internal, external, or global IP address.
+
+required:
+  - ips
+
+properties:
+  ips:
```
Comment (ocsig) [inline community/cloud-foundation/templates/ip_reservation/ip_reservation.py.schema]: I would like to take this into 2 templates.

IP address template: All the logic what is implemented here, except it gets 1 "item" as an input. Properly validates it in the schema, no looping in the template.

IP reservation template: The schema can be exactly what is here with a note to the IP addresses schema. The template itself is only looping through 'ips' and passing each item to the IP Address template.

The goal is to separate the "looping complexity" from the "Multiple type of IP handled uniformly complexity"

File: community/cloud-foundation/templates/ip_reservation/ip_reservation.py.schema [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+      type: object
+      required:
+        - name
+        - ipType
+      properties:
+        name:
+          type: string
+          description: |
+            Name of the reserved IP; unique
+            within the context of the project.
```
Comment (ocsig) [inline community/cloud-foundation/templates/ip_reservation/ip_reservation.py.schema]: - linebreak

File: community/cloud-foundation/templates/ip_reservation/ip_reservation.py.schema [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+        ipType:
+          type: string
+          enum:
+            - global
```
Comment (ocsig) [inline community/cloud-foundation/templates/ip_reservation/ip_reservation.py.schema]: Can we use 'GLOBAL', 'REGIONAL', 'INTERNAL' so no need for additional translation, plus no diversion from the API?

File: community/cloud-foundation/templates/ip_reservation/ip_reservation.py.schema [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+            the specified IP address. If the field is not set, Deployment Manager reserves an internal 
+            IP address that is part of the subnet definition.
+        subnet:
+          type: string
+          description: |
+            [REQUIRED FOR INTERNAL ADDRESSES] The subnet from whose range the IP address is reserved.
+        region:
+          type: string
+          description: |
+            [REQUIRED FOR INTERNAL AND REGIONAL ADDRESSES] ??? Definition ???
```
Comment (ocsig) [inline community/cloud-foundation/templates/ip_reservation/ip_reservation.py.schema]: TODO: Definition

**Comment by tiadobatima1** [comment]
I'll have to loop @sourced-nan here to address the comments. I just submitted the PR while he was on vacation. Cheers :)

File: community/cloud-foundation/templates/ip_reservation/ip_reservation.py.schema (hunk 3 above)
Comment (sourced-vince) [reply community/cloud-foundation/templates/ip_reservation/ip_reservation.py.schema]: The API currently supports "INTERNAL" and "EXTERNAL". Addresses are represented by a compute.v1.address where as global addresses are compute.v1.globalAddress. If this template is to handle both resources, some type of mapping will need to exist or we can add a property field to specify if the address is "GLOBAL", that way we can maintain the `ipType` to just "INTERNAL/EXTERNAL" to match with the API. 

Any preference or suggestions?

Comment (ocsig) [reply community/cloud-foundation/templates/ip_reservation/ip_reservation.py.schema]: I'm okey with the current implementation, just make it all caps so we don't have an internal > INTERNAL translation.

**Comment by sourced-vince** [comment]
Split the templates as suggested into ip_reservation.py and ip_address.py

**Pull request merged**